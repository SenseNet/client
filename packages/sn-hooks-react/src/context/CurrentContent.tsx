import { ConstantContent, ODataParams } from '@sensenet/client-core'
import { GenericContent } from '@sensenet/default-content-types'
import React, { useEffect, useState } from 'react'
import Semaphore from 'semaphore-async-await'
import { useRepository, useRepositoryEvents } from '../hooks'

/**
 * Returns a given content as current content
 */
export const CurrentContentContext = React.createContext<GenericContent>(ConstantContent.PORTAL_ROOT)

export interface CurrentContentProviderProps {
  /**
   * The Id or Path for the current content item
   */
  idOrPath: number | string
  /**
   * Optional callback that will be triggered when the content loads
   */
  onContentLoaded?: (content: GenericContent) => void
  /**
   * Optional OData options for loading the content
   */
  oDataOptions?: ODataParams<GenericContent>
}

/**
 * Provider component for the CurrentContentContext component.
 * Loads a content from the Repository with the given Id or Path.
 * Has to be wrapped with a **RepositoryContext**
 */
export const CurrentContentProvider: React.FunctionComponent<CurrentContentProviderProps> = props => {
  const [loadLock] = useState(new Semaphore(1))
  const [content, setContent] = useState<GenericContent>(ConstantContent.PORTAL_ROOT)
  const repo = useRepository()
  const [reloadToken, setReloadToken] = useState(1)
  const reload = () => setReloadToken(Math.random())
  const events = useRepositoryEvents()

  useEffect(() => {
    const subscriptions = [
      events.onContentModified.subscribe(c => {
        if (c.content.Id === content.Id) {
          reload()
        }
      }),
    ]
    return () => subscriptions.forEach(s => s.dispose())
  }, [content.Id, events.onContentModified])

  const [error, setError] = useState<Error | undefined>()

  useEffect(() => {
    const ac = new AbortController()
    if (props.idOrPath) {
      ;(async () => {
        try {
          const response = await repo.load({
            idOrPath: props.idOrPath,
            requestInit: { signal: ac.signal },
            oDataOptions: props.oDataOptions,
          })
          setContent(response.d)
          props.onContentLoaded && props.onContentLoaded(response.d)
        } catch (err) {
          if (!ac.signal.aborted) {
            setError(err)
          }
        } finally {
          loadLock.release()
        }
      })()
    }
    return () => ac.abort()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [repo, props.idOrPath, reloadToken, loadLock])

  if (error) {
    throw error
  }

  return (
    <CurrentContentContext.Provider value={content as GenericContent}>{props.children}</CurrentContentContext.Provider>
  )
}
