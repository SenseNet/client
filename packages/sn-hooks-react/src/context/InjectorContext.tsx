import { Injector } from '@furystack/inject/dist/Injector'
import React from 'react'

export const InjectorContext = React.createContext(new Injector())
