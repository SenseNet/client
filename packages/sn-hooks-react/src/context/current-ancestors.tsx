import { ODataCollectionResponse } from '@sensenet/client-core'
import { debounce } from '@sensenet/client-utils'
import { GenericContent } from '@sensenet/default-content-types'
import React, { useContext, useEffect, useState } from 'react'
import Semaphore from 'semaphore-async-await'
import { useLogger, useRepository, useRepositoryEvents } from '../hooks'
import { CurrentContentContext } from './current-content'

/**
 * Context that will return with the current content's ancestors
 */
export const CurrentAncestorsContext = React.createContext<GenericContent[]>([])
CurrentAncestorsContext.displayName = 'CurrentAncestorsContext'

export interface CurrentAncestorsProviderProps {
  /**
   * An optional root Path or ID. The top ancestor will be a content with the provided value.
   */
  root?: number | string
}

/**
 * Provider component for the CurrentAncestorsContext component.
 * Loads an ancestor list from the Repository. Has to be wrapped with a **CurrentContentContext** and a **RepositoryContext**
 */
export const CurrentAncestorsProvider: React.FunctionComponent<CurrentAncestorsProviderProps> = (props) => {
  const currentContent = useContext(CurrentContentContext)
  const [loadLock] = useState(new Semaphore(1))

  const [ancestors, setAncestors] = useState<GenericContent[]>([])
  const repo = useRepository()
  const eventHub = useRepositoryEvents()
  const [reloadToken, setReloadToken] = useState(1)

  const logger = useLogger('CurrentAncestorsProvider')

  const requestReload = debounce((newId: number) => {
    if (ancestors.map((a) => a.Id).includes(newId)) {
      setReloadToken(Math.random())
    }
  }, 100)

  useEffect(() => {
    const subscriptions = [
      eventHub.onContentModified.subscribe((mod) => {
        requestReload(mod.content.Id)
      }),
      eventHub.onContentMoved.subscribe((moved) => {
        moved.content.forEach((movedContent) => requestReload(movedContent.Id))
      }),
      eventHub.onContentDeleted.subscribe((del) => {
        del.contentData.forEach((deletedContent) => requestReload(deletedContent.Id))
      }),
    ]
    return () => subscriptions.forEach((s) => s.dispose())
  }, [
    ancestors,
    eventHub.onContentDeleted,
    eventHub.onContentModified,
    eventHub.onContentMoved,
    repo.configuration.repositoryUrl,
    requestReload,
  ])
  const [error, setError] = useState<Error | undefined>()

  useEffect(() => {
    const ac = new AbortController()
    ;(async () => {
      try {
        await loadLock.acquire()
        if ((props.root && currentContent.Id === props.root) || currentContent.Path === props.root) {
          setAncestors([])
        } else if (currentContent.Id) {
          const ancestorsResult = await repo.executeAction<undefined, ODataCollectionResponse<GenericContent>>({
            idOrPath: currentContent.Id,
            method: 'GET',
            name: 'Ancestors',
            body: undefined,
            requestInit: {
              signal: ac.signal,
            },
            oDataOptions: {
              orderby: [['Path', 'asc']],
            },
          })
          const rootIndex = ancestorsResult.d.results.findIndex((a) => a.Id === props.root || a.Path === props.root)
          setAncestors(rootIndex > 0 ? ancestorsResult.d.results.slice(rootIndex) : ancestorsResult.d.results)
        }
      } catch (err) {
        if (!ac.signal.aborted) {
          setError(err)
        }
      } finally {
        loadLock.release()
      }
    })()
    return () => ac.abort()
  }, [currentContent.Id, currentContent.Path, loadLock, props.root, reloadToken, repo])

  if (error) {
    logger.warning({
      message: `Error loading ancestors. ${error.message}`,
      data: { error, relatedContent: currentContent, relatedRepository: repo.configuration.repositoryUrl },
    })
  }
  return <CurrentAncestorsContext.Provider value={ancestors}>{props.children}</CurrentAncestorsContext.Provider>
}
