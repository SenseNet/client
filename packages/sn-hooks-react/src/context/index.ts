export * from './current-ancestors'
export * from './current-children'
export * from './current-content'
export * from './injector'
export * from './load-settings'
export * from './logger'
export * from './repository'
export * from './session'
