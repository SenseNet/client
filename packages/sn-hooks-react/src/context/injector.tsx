import { Injector } from '@sensenet/client-utils'
import { createContext } from 'react'

/**
 * Context that returns an Injector instance
 */
export const InjectorContext = createContext(new Injector())
InjectorContext.displayName = 'InjectorContext'
