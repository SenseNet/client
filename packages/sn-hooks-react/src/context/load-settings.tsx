import { ODataParams } from '@sensenet/client-core'
import { GenericContent } from '@sensenet/default-content-types'
import React, { Dispatch } from 'react'

export interface LoadSettingsContextProps {
  /**
   * OData settings for the Current Content
   */
  loadSettings?: ODataParams<GenericContent>
  /**
   * OData settings for loading the current content's children
   */
  loadChildrenSettings: ODataParams<GenericContent>
  /**
   * OData settings for loading ancestors
   */
  loadAncestorsSettings?: ODataParams<GenericContent>

  /**
   * Sets the Current Content's load settings
   */
  setLoadSettings: Dispatch<ODataParams<GenericContent>>

  /**
   * Set the current content's childrens load settings
   */
  setLoadChildrenSettings: Dispatch<ODataParams<GenericContent>>

  /**
   * Sets the current content's ancestor's load settings
   */
  setLoadAncestorsSettings: Dispatch<ODataParams<GenericContent>>
}

/**
 * Context that stores load settings for OData requests
 */
export const LoadSettingsContext = React.createContext<LoadSettingsContextProps | undefined>(undefined)

const initialLoadChildrenSettings: ODataParams<GenericContent> = {
  orderby: [['DisplayName', 'asc']],
  select: 'all',
  expand: ['CreatedBy'],
}
/**
 * Provider for the LoadSettingsContext. Sets up default loading settings.
 */

const setLoadSettingsReducer = (
  state: ODataParams<GenericContent> | undefined,
  value: ODataParams<GenericContent> | undefined,
) => {
  if (JSON.stringify(value) === JSON.stringify(state)) {
    return state
  }
  return value
}

const setLoadChildrenSettingsReducer = (state: ODataParams<GenericContent>, value: ODataParams<GenericContent>) => {
  if (JSON.stringify(value) === JSON.stringify(state)) {
    return state
  }
  return value
}

export const LoadSettingsContextProvider: React.FunctionComponent = props => {
  const [loadSettings, setLoadSettings] = React.useReducer(setLoadSettingsReducer, undefined)
  const [loadChildrenSettings, setLoadChildrenSettings] = React.useReducer(
    setLoadChildrenSettingsReducer,
    initialLoadChildrenSettings,
  )
  const [loadAncestorsSettings, setLoadAncestorsSettings] = React.useReducer(setLoadSettingsReducer, undefined)

  return (
    <LoadSettingsContext.Provider
      value={{
        loadSettings,
        loadChildrenSettings,
        loadAncestorsSettings,
        setLoadSettings,
        setLoadChildrenSettings,
        setLoadAncestorsSettings,
      }}>
      {props.children}
    </LoadSettingsContext.Provider>
  )
}
