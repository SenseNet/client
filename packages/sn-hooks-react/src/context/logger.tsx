import { Logger, LoggerCollection } from '@sensenet/client-utils'
import React, { createContext, FunctionComponent } from 'react'
import { useInjector } from '../hooks'
/**
 * Context for a Logger instance
 */
export const LoggerContext = createContext<Logger>(new LoggerCollection())
LoggerContext.displayName = 'LoggerContext'

/**
 * Wrapper for the LoggerContext component.
 * Returns the current injector's default logger. Has to be wrapped with **InjectorContext**
 */
export const LoggerContextProvider: FunctionComponent = ({ children }) => {
  const injector = useInjector()
  return <LoggerContext.Provider value={injector.logger}>{children}</LoggerContext.Provider>
}
