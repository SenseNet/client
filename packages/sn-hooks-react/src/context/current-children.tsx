import { ODataParams } from '@sensenet/client-core'
import { deepMerge, PathHelper } from '@sensenet/client-utils'
import { GenericContent } from '@sensenet/default-content-types'
import { Created } from '@sensenet/repository-events'
import React, { useContext, useEffect, useState } from 'react'
import { useRepository, useRepositoryEvents } from '../hooks'
import { CurrentContentContext } from './current-content'
import { LoadSettingsContext } from './load-settings'

/**
 * Context that will return with a list of a current content's children
 */
export const CurrentChildrenContext = React.createContext<GenericContent[]>([])
CurrentChildrenContext.displayName = 'CurrentChildrenContext'

export interface CurrentChildrenProviderProps {
  loadSettings?: ODataParams<GenericContent>
}

/**
 * Provider component for the CurrentChildrenContext component
 * Loads the children of the current content.
 * Loads an ancestor list from the Repository. Has to be wrapped with a **CurrentContentContext** and a **RepositoryContext**
 */
export const CurrentChildrenProvider: React.FunctionComponent<CurrentChildrenProviderProps> = (props) => {
  const currentContent = useContext(CurrentContentContext)
  const [children, setChildren] = useState<GenericContent[]>([])

  const [reloadToken, setReloadToken] = useState(1)
  const repo = useRepository()
  const eventHub = useRepositoryEvents()
  const loadSettings = useContext(LoadSettingsContext)

  const requestReload = () => setReloadToken(Math.random())
  const [error, setError] = useState<Error | undefined>()

  useEffect(() => {
    const ac = new AbortController()
    ;(async () => {
      if (currentContent.Path) {
        try {
          const childrenResult = await repo.loadCollection<GenericContent>({
            path: currentContent.Path,
            requestInit: { signal: ac.signal },
            oDataOptions: deepMerge(loadSettings.loadChildrenSettings, props.loadSettings),
          })
          setChildren(childrenResult.d.results)
        } catch (err) {
          if (!ac.signal.aborted) {
            setError(err)
          }
        }
      }
    })()
    return () => ac.abort()
  }, [currentContent.Path, loadSettings.loadChildrenSettings, props.loadSettings, repo, reloadToken])

  useEffect(() => {
    const handleCreate = (c: Created) => {
      if ((c.content as GenericContent).ParentId === currentContent.Id) {
        return requestReload()
      }
      if (PathHelper.isAncestorOf(currentContent.Path, c.content.Path)) {
        requestReload()
      }
    }

    const subscriptions = [
      eventHub.onCustomActionExecuted.subscribe((event) => {
        if (event.actionOptions.method !== 'GET') {
          requestReload()
        }
      }),
      eventHub.onContentCreated.subscribe(handleCreate),
      eventHub.onContentCopied.subscribe(handleCreate),
      eventHub.onContentMoved.subscribe(handleCreate),
      eventHub.onContentModified.subscribe((mod) => {
        if (mod.forceRefresh || children.some((c) => c.Id === mod.content.Id)) {
          requestReload()
        }
      }),

      eventHub.onUploadFinished.subscribe((data) => {
        if (PathHelper.isAncestorOf(currentContent.Path, data.Url)) {
          requestReload()
        }
      }),
      eventHub.onContentDeleted.subscribe((d) => {
        if (PathHelper.getParentPath(d.contentData.Path) === PathHelper.trimSlashes(currentContent.Path)) {
          requestReload()
        }
      }),
    ]

    return () => subscriptions.forEach((s) => s.dispose())
  }, [
    currentContent,
    repo,
    children,
    eventHub.onCustomActionExecuted,
    eventHub.onContentCreated,
    eventHub.onContentCopied,
    eventHub.onContentMoved,
    eventHub.onContentModified,
    eventHub.onContentDeleted,
    eventHub.onUploadFinished,
  ])

  if (error) {
    throw error
  }

  return <CurrentChildrenContext.Provider value={children}>{props.children}</CurrentChildrenContext.Provider>
}
