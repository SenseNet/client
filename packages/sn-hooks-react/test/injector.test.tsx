import { shallow } from 'enzyme'
import React from 'react'
import { InjectorContext } from '../src/context/injector'

describe('Injector', () => {
  it('matches snapshot', () => {
    const p = shallow(<InjectorContext.Consumer>{() => <></>}</InjectorContext.Consumer>)
    expect(p).toMatchSnapshot()
  })
})
