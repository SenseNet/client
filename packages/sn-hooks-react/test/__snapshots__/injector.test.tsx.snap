// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Injector matches snapshot 1`] = `<Fragment />`;
